Sources/StreamChat/Controllers/SearchControllers/UserSearchController/UserSearchController.swift

```diff
@@ -195,16 +96,129 @@ public class ChatUserSearchController: DataController, DelegateCallable, DataSto
         limit: Int = 25,
         completion: ((Error?) -> Void)? = nil
     ) {
-        guard let lastQuery = lastQuery else {
+        guard let lastQuery = query else {
             completion?(ClientError("You should make a search before calling for next page."))
             return
         }
         
         var updatedQuery = lastQuery
-        updatedQuery.pagination = Pagination(pageSize: limit, offset: users.count)
-        userQueryUpdater.update(userListQuery: updatedQuery) { error in
-            self.callback { completion?(error) }
+        updatedQuery.pagination = Pagination(pageSize: limit, offset: userArray.count)
+        
+        fetch(updatedQuery, completion: completion)
+    }
+}
+
+private extension ChatUserSearchController {
+    /// Fetches the given query from the API, saves the loaded page to the database, updates the list of users and notifies the delegate.
+    ///
+    /// - Parameters:
+    ///   - query: The query to fetch.
+    ///   - completion: The completion that is triggered when the query is processed.
+    func fetch(_ query: UserListQuery, completion: ((Error?) -> Void)? = nil) {
+        // TODO: Remove with the next major
+        //
+        // This is needed to make the delegate fire about state changes at the same time with the same
+        // values as it was when query was persisted.
+        setLocalDataFetchedStateIfNeeded()
+        
+        userQueryUpdater.fetch(userListQuery: query) { result in
+            switch result {
+            case let .success(page):
+                self.save(page: page) { loadedUsers in
```

Comment:
Possible memory leak

Reply:
Whilst those should not be leaks, I will address your comments for now. We might need to add some conversation wrt these. Weakifying everywhere is "fine", but also adds some overhead

Reply:
It is safer just to default weakify every closure. Especially in an SDK, we can't afford to have crashes or memory leaks, so we should try to do our best to avoid this. Also it future proofs possible refactorings, since right now it might not leak, but if some refactoring happens in the Updater or something like this, that could change.